src/runtime_src/core/edge/skd/xrt_skd.cpp

```diff
@@ -35,313 +40,370 @@ namespace xrt {
+
+    while (true) {
+      ret = wait_next_cmd();
+      cmd_start = Clock::now();
+      if(cmd_end < cmd_start) {
+	  std::string msg = std::string("PS Kernel Command interval = ") + std::to_string((std::chrono::duration_cast<ms_t>(cmd_start - cmd_end)).count());
+	  xrt_core::message::send(severity_level::info, "SKD", msg);
       }
 
-      syslog(LOG_INFO, "Got new kernel command!\n");
+      if (ret && (signal==SIGTERM)) {
+	  // We are told to exit the soft kernel loop
+	  std::string msg = std::string("Exit soft kernel %s\n") + m_sk_name;
+	  xrt_core::message::send(severity_level::info, "SKD", msg);
+	  break;
+      }
 
-      /* Reg file indicates the kernel should not be running. */
-      if (!(args_from_host[0] & 0x1))
+      // Reg file indicates the kernel should not be running.
+      if (!(m_args_from_host[0] & 0x1))
 	continue; //AP_START bit is not set; New Cmd is not available
 
       // FFI PS Kernel implementation
       // Map buffers used by kernel
-      for(int i=0;i<num_args;i++) {
-	if((args[i].index == xrt_core::xclbin::kernel_argument::no_index) && (args[i].hosttype.compare("xrtHandles*")==0)) {
-	  ffi_arg_values[i] = &xrtHandle;
-	} else if(args[i].type == xrt_core::xclbin::kernel_argument::argtype::global) {
-	  uint64_t *buf_addr_ptr = (uint64_t*)(&args_from_host[(args[i].offset+PS_KERNEL_REG_OFFSET)/4]);
-	  uint64_t buf_addr = reinterpret_cast<uint64_t>(*buf_addr_ptr);
-	  uint64_t *buf_size_ptr = (uint64_t*)(&args_from_host[(args[i].offset+PS_KERNEL_REG_OFFSET)/4+2]);
-	  uint64_t buf_size = reinterpret_cast<uint64_t>(*buf_size_ptr);
-	  boSize[i] = buf_size;
-
-	  boHandles[i] = xclGetHostBO(devHdl,buf_addr,buf_size);
-	  bos[i] = xclMapBO(devHdl,boHandles[i],true);
-	  ffi_arg_values[i] = &bos[i];
+      for(int i=0;i<m_kernel_args.size();i++) {
+	if((m_kernel_args[i].index == xrt_core::xclbin::kernel_argument::no_index) && (m_kernel_args[i].hosttype.compare("xrtHandles*")==0)) {
+	  ffi_arg_values[i] = &m_xrtHandle;
+	} else if(m_kernel_args[i].type == xrt_core::xclbin::kernel_argument::argtype::global) {
+	  auto buf_addr_ptr = reinterpret_cast<uint64_t *>(&m_args_from_host[(m_kernel_args[i].offset + PS_KERNEL_REG_OFFSET) / 4]);
+	  auto buf_addr = reinterpret_cast<uint64_t>(*buf_addr_ptr);
+	  auto buf_size_ptr = reinterpret_cast<uint64_t *>(&m_args_from_host[(m_kernel_args[i].offset + PS_KERNEL_REG_OFFSET) / 4 + 2]);
+	  auto buf_size = reinterpret_cast<uint64_t>(*buf_size_ptr);
+
+#ifdef SKD_MAP_BIG_BO
+	  bo_offsets[i] = buf_addr - mem_start_paddr;
+	  bos[i] = mem_start_vaddr + bo_offsets[i];
+#else
+	  bo_size[i] = buf_size;
+	  bo_handles[i] = xclGetHostBO(m_devhdl, buf_addr, buf_size);
+	  bos[i] = xclMapBO(m_devhdl, bo_handles[i], true);
 	  bo_list.emplace_back(i);
+#endif
+	  std::stringstream strm;
+	  strm << bos[i];
+	  std::string msg = std::string("BO Virtual Address = ") + strm.str();
+	  xrt_core::message::send(severity_level::debug, "SKD", msg);
+	  ffi_arg_values[i] = &bos[i];
```

Comment:
Cross dependent binding.  
This could lead to future issues if the bos vector collection array changes.  This can happen when the vector needs to increase its side.  When this occurs a new array is created, the old array is copied to the new array, and the old array removed from memory.